docs/RFCS/20191009_gc_protected_timestamps.md

```diff
@@ -0,0 +1,597 @@
+  Protect(
+    ctx context.Context,
+    txn *client.Txn,
+    ts hlc.Timestamp,
+    mode ProtectionMode,
+    metaType string, meta []byte,
+    spans ...roachpb.Span,
+  ) (uuid.UUID, error)
+
+  // Release allows spans which were previously protected to now be garbage
+  // collected.
+  //
+  // If the specified UUID does not exist ErrNotFound is returned but the
+  // passed txn remains safe for future use.
+  Release(_ context.Context, txn *client.Txn, id uuid.UUID) error
+}
+```
+
+### `protectedtsbase`
+
+```go
+package protectedts
+
+type Record struct {
+  ID        uuid.UUID
+  Timestamp hlc.Timestamp
+  Mode      ProtectionMode
+  MetaType  string
+  Meta      []byte
+  Spans     []roachpb.Span
+}
+
+// Tracker will be used in the storage package to determine a safe
+// timestamp for garbage collection.
+type Tracker interface {
+  // TODO(ajwerner): Is there anything useful to do with an error?
+  // Should this just return records and a timestamp and if failed then
+  // return a timestamp of 0 and log internally?
+  GetProtectedTimestamps(intersecting roachpb.Span) ([]Record, hlc.Timestamp, error)
+  
+  // TODO(ajwerner): consider providing a method to force refreshing the
+  // state as of the current time. Clients may want to call this when
+  // zone configs are changing. 
+}
+```
+
+### `protectedts.Catalog`
+
+State reconciliation can be implemented by retrieving the entire state
+of the protected timestamp subsystem. The complete state looks like the
+following:
+
+```go 
+package protectedts
+
+type State struct {
+  Metadata
+  Records []protectedtsbase.Record
+}
+
+type Metadata struct {
+  Version    int
+  NumRecords int
+  NumSpans   int
+}
+
+type Catalog interface {
+  // List returns the complete state of the protected timestamp subsystem.
+  List(context.Context, *client.Txn) (*State, error)
+
+  // GetMetadata can be used to cheaply update the timestamp at
+  // which a listing was "observed". All changes to State will
+  // increment the version so if a later call to GetMetadata
+  // returns the same version as an earlier call to List then
+  // it is guaranteed that the value returned from List at that
+  // timestamp would not have changed.
+  GetMetadata(context.Context, *client.Txn) (*Metadata, error)
+}
+```
+
+### Additional Changes to storage
+
+Each replica currently keeps an in-memory copy of its zone config.
+This value is updated based on gossip. In the gossiped system config
+we already include MVCC timestamps. In order to implement the GC protocol
+the Replica needs to be aware of the MVCC timestamp at which its zone config
+was written.
+
+### Changes to Zone Configuration Updates
+
+Zone configuration updates need to observe the protected timestamp
+metadata. This is important so that when protected timestamps are written
+they know that they have seen an up-to-date copy of the zone config. We do
+this by relying on the single-key linearizability offered by CockroachDB.
```

Comment:
> relying on the single-key linearizability offered by CockroachDB

This is only guaranteed for a read after a write if the system is within clock offset bounds. Meanwhile, single-key linearizability is guaranteed between two writes regardless of clock skew. If we expect zone cfg updates to be infrequent then it may be worth performing an identity write on the protected timestamp metadata to eliminate the clock dependency.